consensus/src/runway/packer.rs

```diff
@@ -0,0 +1,193 @@
+                break;
+            }
+        }
+    }
+
+    pub async fn run(&mut self, mut exit: oneshot::Receiver<()>) -> Result<(), ()> {
+        info!(target: "AlephBFT-packer", "{:?} Packer started.", self.index());
+        let pack = self.pack().fuse();
+        pin_mut!(pack);
+        futures::select! {
+            _ = pack => Err(()),
+            _ = exit => Ok(()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Packer;
+    use crate::{
+        units::{ControlHash, PreUnit, SignedUnit},
+        NodeCount, NodeIndex, Receiver, Sender, SessionId,
+    };
+    use aleph_bft_mock::{Data, DataProvider, Hasher64, Keychain};
+    use aleph_bft_types::NodeMap;
+    use futures::{
+        channel::{mpsc, oneshot},
+        pin_mut, FutureExt, StreamExt,
+    };
+
+    const SESSION_ID: SessionId = 43;
+    const NODE_ID: NodeIndex = NodeIndex(0);
+    const N_MEMBERS: NodeCount = NodeCount(4);
+
+    fn prepare(
+        keychain: &Keychain,
+    ) -> (
+        Sender<PreUnit<Hasher64>>,
+        Receiver<SignedUnit<Hasher64, Data, Keychain>>,
+        Packer<Hasher64, Data, DataProvider, Keychain>,
+        oneshot::Sender<()>,
+        oneshot::Receiver<()>,
+        PreUnit<Hasher64>,
+    ) {
+        let data_provider = DataProvider::new();
+        let (preunits_channel, preunits_from_runway) = mpsc::unbounded();
+        let (signed_units_for_runway, signed_units_channel) = mpsc::unbounded();
+        let packer = Packer::new(
+            data_provider,
+            preunits_from_runway,
+            signed_units_for_runway,
+            keychain,
+            SESSION_ID,
+        );
+        let (exit_tx, exit_rx) = oneshot::channel();
+        let parent_map = NodeMap::with_size(N_MEMBERS);
+        let control_hash = ControlHash::new(&parent_map);
+        let preunit = PreUnit::new(NODE_ID, 0, control_hash);
+        (
+            preunits_channel,
+            signed_units_channel,
+            packer,
+            exit_tx,
+            exit_rx,
+            preunit,
+        )
+    }
+
+    #[tokio::test]
+    async fn unit_packed() {
+        let keychain = Keychain::new(N_MEMBERS, NODE_ID);
+        let (preunits_channel, signed_units_channel, mut packer, _exit_tx, exit_rx, preunit) =
+            prepare(&keychain);
+        let packer_handle = packer.run(exit_rx).fuse();
+        preunits_channel.unbounded_send(preunit.clone()).unwrap();
+        pin_mut!(packer_handle);
+        pin_mut!(signed_units_channel);
+        let unit = futures::select! {
+            unit = signed_units_channel.next() => match unit {
+                Some(unit) => unit,
+                None => panic!(),
+            },
+            _ = packer_handle => panic!(),
+        }
+        .into_unchecked()
+        .into_signable();
+        assert_eq!(SESSION_ID, unit.session_id());
+        assert_eq!(unit.as_pre_unit(), &preunit);
+    }
+
+    #[tokio::test]
+    async fn closed() {
```

Comment:
I don't get the purpose of this test.

Reply:
It either requires a better name or an explanation in a comment.

Reply:
renamed

Reply:
I still don't get it. D'''x